### PR TITLE
add special handling for M220 and M221

### LINKED ISSUE #192: Wrong estimated time because of wrong printer_config
OctoPrint Version: 1.4.0
PrintTimeGenius Version: 2.2.1

PrintTimeGenius calculated wrong printing times (way too high or low).
After searching for the error I found that `printer_config.yaml` contained wrong values. After deleting the wrong values and deleting `print_history.yaml` (to prevent wrong compensation of the new analysis) PrintTimeGenius calculates the correct print time again.

The wrong values in `printer_config.yaml` were the following:
```
M220 S50      // set feedrate to 50%
M200 D1.75    // set filament diameter to 1.75mm
M200 D0       // set filament diameter to 0mm (disable volumetric extrusion)
```

So there is must be a bug in `printer_config.py` because the two mcodes M200 should have been merged because M and T match (mcode is the same and T is present for both mcodes) and D is present for both mcodes (see comment in `printer_config.py`).

Also I didn't find these mcodes in any uploaded gcode file (and it should not be in any gcode file). PrintTimeGenius listens for commands sent to the printer and always updates the printer config when it finds a matching mcode. And that is a problem. When the user manually changes the feedrate (or any other non-persistent settings) in the OctoPrint UI, PrintTimeGenius saves this new value. After powering the printer off and on it's feedrate (or the other non-persistent settings) is reset to 100% but PrintTimeGenius will calculate print times based on wrong values which leads to wrong printing times at the beginning and bad compensated printing times later.

### PR DESCRIPTION
**Pull request by jonnykl**
partially fixes #192

### CODE REVIEW

**Comment by jonnykl** [comment]
Only tested with single extruder but should also work with multiple extruders

**Comment by eyal0** [comment]
That looks pretty good. You tested it and found that it was correctly adding those M commands into the printer config yaml?

**Comment by eyal0** [comment]
Can you check that the M220 and M221 are affecting the Marlin calculation?

You can do that by uploading a new file or by running marlin-calc directly with arguments.

**Comment by eyal0** [comment]
Looks like Marlin calc does support those:

https://github.com/eyal0/Marlin/blob/e7e9ca357372c78f94bc8a52e72d4f99c2116787/Marlin/src/module/calc.cpp#L449

So it ought to work.

**Comment by jonnykl** [comment]
> That looks pretty good. You tested it and found that it was correctly adding those M commands into the printer config yaml?

Yes, they appear in the printer config yaml. Now also tested with multi extruder setup.

> Can you check that the M220 and M221 are affecting the Marlin calculation?
> 
> You can do that by uploading a new file or by running marlin-calc directly with arguments.

Yes, I tested it by running marlin-calc directly and it can make a huge difference (e.g. when using a feed rate of 50% vs 100%).

File: octoprint_PrintTimeGenius/printer_config.py [hunk 1]
```diff
@@ -134,6 +134,10 @@ def __iadd__(self, new_line):
     mcode = get_code(new_line, "M")
     if not mcode:
       return self  #Save time by quiting early
+    # add T0 when extruder index is ommited to correctly merge the mcodes
+    if mcode == "M221":
+      if get_code(new_line, "T") == "":
+        new_line += " T0"
```
Comment (eyal0) [inline octoprint_PrintTimeGenius/printer_config.py:140]: Want to add some doctest around line 125?

https://docs.python.org/3/library/doctest.html

**Pull request merged**